bitcoin/src/bip32.rs

```diff
@@ -114,16 +114,10 @@ internals::serde_string_impl!(Xpub, "a BIP-32 extended public key");
 /// A child number for a derived key
 #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
 pub enum ChildNumber {
-    /// Non-hardened key
-    Normal {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
-    /// Hardened key
-    Hardened {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
+    /// A normal child number, indexes 0 to 2147483647 (the first half of all possible children)
+    Normal(u32),
+    /// A hardened child number, indexes 2147483648 to 4294967295 (the second half of all possible children)
+    Hardened(u32),
```

Comment:
These need to be newtypes.

Reply:
Why ? Please be clear while requesting changes.
The code deals great with them being enums.

Reply:
We've decided this elsewhere. It models the situation much better because not all values are valid. The current code allows people to create garbage child numbers.